### ISSUE TITLE
is fairscale usefull for small model but high memory consumption in the forward pass

### ISSUE THREAD

**Issue by lolongcovas** [facebookresearch/fairscale#628]
## ❓ Questions and Help
Hi,

I was training an efficientdet [with efficientnet-b2 as backbone] with 4 gtx-1070 8 gb each. the batch size that fits, in each gpu, is 4. This amount of bsize is not good for learning the batch norm params, then I decide to use fairscale to save the memory usage and increase the batch size. After looking at the model parameters, the weights are only 35Mb aprox. I guess the high gpu memory usage are due to the layer output feature maps. According to my understanding, the faircale toolbox does model weight shard optimization, in my case 35Mb will not help. isnt it? If i am wrong, could u guys suggest me the correct package to use for increasing the batch size?

**Comment by visionscaper** [comment]
Hi @lolongcovas, if you have a lot of activation data, you could use activation checkpoints. See Pytorch activation checkpoints https://pytorch.org/docs/stable/checkpoint.html

Fairscale also has its own version https://github.com/facebookresearch/fairscale/blob/master/fairscale/nn/misc/checkpoint_activations.py

Maybe the Fairscale team can tell you more about that.

**Comment by lolongcovas** [comment]
Hi @visionscaper thanks for the response, I also thought that maybe a solution. I am going to take a look the checkpoint. thanks!

**Comment by min-xu-ai** [comment]
Thanks @visionscaper! You are absolutely right. More documentation of `checkpoint_wrapper` is here:

https://fairscale.readthedocs.io/en/latest/api/nn/misc/checkpoint_activations.html

You can find more examples of how we use it in the `tests` dir. @lolongcovas.

For batchnorm, because it tracks the running stats during the forward pass, the pytorch version of checkpoint will not automatically work with batchnorms. I recommend you using `checkpoint_wrapper` because it will "patch" your batchnorm classes and ensure that the running stats are not updated incorrectly in the forward pass.

Deciding how to split your model into different checkpointed parts is not always trivial. You may want to try a few ways to split your model to see which way saves the most memory for you. It is hard to prove that you have gotten the optimal memory saving but in my experiences getting a solution that saves "enough" memory is not that hard.

By "splitting", I mean something like:

model = [layer1, layer2, layer3, ... layer n]

You can try to split it 2, 3, 4, ... parts. Say, you want to try split==2 first, then you can try to split in the middle:

model = checkpoint(layer1, layer2, .. layer n/2), checkpoint(layer n/2+1, ... layer n)

You can also try to split based on parameter size or activation size etc. 

Finally, different pytorch version may give you different memory saving results due to (I suspect) different behavior of the memory allocators. You may want to try different versions if you are not seeing good memory savings.

**Comment by min-xu-ai** [comment]
I should have mentioned that fairscale also has a few methods for offloading the activation tensors to CPU memory. So, if you have a lot of CPU memory to spare, we can discuss that in more details as well. Just let us know.

**Comment by lolongcovas** [comment]
thanks, I will look at into it.
